### ISSUE TITLE
Remote Uri are extremely difficult to handle correctly as an extension dev

### ISSUE THREAD

**Issue by stevenguh** [microsoft/vscode#105969]
Split from https://github.com/microsoft/vscode/issues/82739#issuecomment-683330593

Path manipulation as an extension developer seems to be extremely difficult to handle correctly especially accounting the vscode-remote's Uri (without local installation on remote). The issue mostly comes from the early separator normalization base on the host platform in the vscode's API.

I have been somewhat involved in handling paths in a few open-source projects, and here are some API issues that I ran into for cases that a users remoting into other platform without installing that extension locally.

1. `Uri.fsPath`:
- Return path with the host's separator which means
    + `C:/path/to/file` is returned if *nix remoting into a Windows
    + `\path\to\file` is returned if Windows machine remoting into *nix
    + Normalization to Windows path (`\` separator, *nix remoting  into Windows) cause any path that has `\` in the *nix remote indistinguishable from separator. Hence, force developer to use other properties like `Uri.path` to handle the case when a path contains `\` character in the *nix remote.
https://github.com/microsoft/vscode/blob/f74e473238aca7b79c08be761d99a0232838ca4c/src/vs/base/common/uri.ts#L600-L602

2. `workspace.asRelativePath`:
- Always return `/` as separator if the input uri is in any of the workspace folders.
https://github.com/microsoft/vscode/blob/f74e473238aca7b79c08be761d99a0232838ca4c/src/vs/base/common/resources.ts#L232-L235
- Else return `Uri.fsPath` of the input uri
    + The normalization of `Uri.fsPath` means that it will return path with `\` on Windows machine
    + The inconsistent separator will make this API difficult to use and as well as to handle the edge case I mentioned above

3. Lack of API to detect if the remote environment is Windows or *nix

Here are the issues I've been involved:
- https://github.com/VSpaceCode/VSpaceCode/issues/83
    + This is an issue to implement commands to copy different type of paths (absolute, relative, w/ & w/o line and column) of the currently active editor into clipboard.
    + I handled it by reimplementing some of the vscode apis without the platform based path normalization and normalized it at the end by guessing the platform with the format of the Uri
    + See [here]( https://github.com/VSpaceCode/VSpaceCode/blob/ff26627a645e95faaf50dc5b4c65f49b6b9140c5/src/pathCommands.ts) for the implementation
- https://github.com/VSCodeVim/Vim/issues/5158
    + This is still a WIP to refactor and fix some of the assumption I made (e.g. remote server can only be *nix) of feature to open file with freeform text. This is case where it's tough to handle path joining with freeform text without knowing the remote platform. (especially `\` is not a path separator on *nix)
- https://github.com/bodil/vscode-file-browser/issues/6
    + The author decided to make a compromise to display with an incorrect path separator when remoting into a different platform and other edge cases.

This is an issue with multiple fronts, and let me know if any of it is confusing. As an extension developer, I wish I didn't have to learn all the inner working to work with path correctly (in all platforms and vscode-remote).

**Comment by jrieken** [comment]
Not quite sure what outcome is expected from this issue. Generally, my advice is not to use fspath (which we should consider to deprecate) and always use the uri. Is this a display issue?

**Comment by stevenguh** [comment]
Sorry for not articulating well with the expected outcome. Let me try again.

1. Display is definitely one of the pain points. If `fsPath` is deprecated, there should be another api so that an extension developer can display Uri correctly with the right separator without local extension installation (for both local file path and remote path).
2. `workspace.asRelativePath` returns inconsistent path separator due its usage of `fsPath` as mentioned above. This should ideally be fixed, so the extension doesn't have to reimplement the internally API differently just to get around this ([example](https://github.com/VSpaceCode/VSpaceCode/blob/ff26627a645e95faaf50dc5b4c65f49b6b9140c5/src/pathCommands.ts#L109-L123)).
3. Even working on `Uri.path`, windows path can still be difficult to handle
   - If I have an extension that a user can traverse the `uri.path` for example, then I will need to know when to to stop traversing up the path i.e. if  `Uri.path` is `/c:/`. I can do that with with local file because I can access the platform information (e.g. via node's path api); however, extension do not have access to platform info with `Uri` of scheme  `vscode-remote`. This means that an extension either let the user traverse all the way up to `/` or guess the platform based on the `Uri`.
    - Guessing platform base on the the drive letter on Uri would have an edge case that `/c:/` on linux will be assumed to be Windows. In addition, guessing will fail to guess the platform if the path is a user generated one (e.g. `/Python38/` is handled by `workspace.fs` as `/c:/Python38/` and it would been guessed as *nix path)
   - In short, `Uri` alone cannot solve the question of should `\` be handle as a separator from user input and what's the possible root of the path for that platform.
   - This is where I don't know what's the right solution for this should be especially vscode internally seems have a better way to handle Uri. Exposing the platform of the remote will allow extension develop to handle this.
4. Another additional pain point, there is no API to get the `Uri` if the user didn't open any folder and file while remote in with `vscode-remote` (AFAIK anyway)

For the most of the display issue, maybe exposing resourceLabelFormatters api would help? So extension can handle Uri correctly for display purposes.

Thank you for reading such a long explanation. I hope this clarifies somewhat. The issue seems multitudinal, which makes it hard to explain.

**Comment by jrieken** [comment]
> Display is definitely one of the pain points. 

How/where do you display uris or paths? Should should mostly be taken care of by VS Code itself

**Comment by stevenguh** [comment]
VSpaceCode/VSpaceCode#83:
This get the abs/relative path of the active editor, copy it to the clipboard and display it on the stats bar.
![image](https://user-images.githubusercontent.com/3455662/92162284-c74a0780-ede6-11ea-8f73-508987dbc425.png)

VSCodeVim/Vim#5158:
This is emulating the `:e` file opening in Vim, where a user can enter a free from text and tabbing it would allow autocompletion and it's displayed on status bar.
![image](https://user-images.githubusercontent.com/3455662/92162328-db8e0480-ede6-11ea-808b-2c2104a70d5b.png)

bodil/vscode-file-browser#6:
This is similar to file opening dialog in vscode-remote. It uses QuickPick API to display the current path as the title.
![image](https://user-images.githubusercontent.com/3455662/92162381-efd20180-ede6-11ea-80d4-fcff76cc17d9.png)

Can you elaborate a little bit on "being taken care by VS Code"?

**Comment by jrieken** [comment]
Thanks for providing those samples. Internally, VS Code knows how to represent an uri best, e.g absolute or not, what path separators, custom rules, etc. Unsure if we can and should expose those APIs or if we should simply accept an uri is more places, e.g the text of a status bar item could be a uri and we would take care of the formatting etc pp

**Comment by stevenguh** [comment]
As far as for the display problem, I would say the former to enable greater flexibly for an extension developer. There is gonna be usages that need a formatted uri but doesn't display in the vscode's selected ui. I would imagine it being like an API that an extension developer can pass an uri. (I assume this is related to how vscode internally uses `contributes.resourceLabelFormatters`)

Also, knowing the separator is important to handle the vim open file command, otherwise, the best we can do is to either guess the separator base on the uri or use the host platform's separator.

**Comment by jrieken** [comment]
Adding @isidorn who owns the label formatter

**Comment by isidorn** [comment]
Currently there is a proposed API for extensions to contribute ResourceLabelFormatters that control how URIs are displayed across the workbench. Code pointer https://github.com/microsoft/vscode/blob/82aa8fc425d976cceed8fc175a388f4b8e9ce640/src/vs/vscode.proposed.d.ts#L267
However I am not sure if we will finalize this API since there were no big requests from the community thus far.

As for having a method for extensions to pass a URI and we return the display value, that is a fair API feature request which we might add in the future. Feel free to create a new issue for this.

**Comment by jrieken** [comment]
> Feel free to create a new issue for this.

I'd say this is the issue for that

**Comment by jrieken** [comment]
related to https://github.com/microsoft/vscode/issues/108256
